From a438d4f9922bd5b33672f97c0f971c5e64d7c180 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Wed, 14 Nov 2018 14:23:48 +0200
Subject: [PATCH 0791/1051] net: mvpp2: move xps setting into port-probe

Netdev multi tx queue XPS setting should be done before netdev
registration.
Late mapping may work improperly, and also the read
over file /sys/class/net/eth<n>/queues/tx-<n>/xps_cpus may be
incorrect.

So moving the XPS setting from start-dev into port-probe.

For the MVPP2 logic, do not map several queues onto same CPU.
Unmapped queues are an "extension" for Rx-to-Tx CoS traffic.

Change-Id: Id80e1054b801a2ea600a7f2f9f870059c6f827ea
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/60915
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 14 +++++---------
 1 file changed, 5 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index 3e2f1c6f99d0..e47619a22f12 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -2617,23 +2617,15 @@ static int mvpp2_setup_rxqs(struct mvpp2_port *port)
 static int mvpp2_setup_txqs(struct mvpp2_port *port)
 {
 	struct mvpp2_tx_queue *txq;
-	int queue, err, cpu;
+	int queue, err;
 
 	for (queue = 0; queue < port->ntxqs; queue++) {
 		txq = port->txqs[queue];
 		err = mvpp2_txq_init(port, txq);
 		if (err)
 			goto err_cleanup;
-
-		/* Assign this queue to a CPU */
-		cpu = queue % num_present_cpus();
-		netif_set_xps_queue(port->dev, cpumask_of(cpu), queue);
 	}
 
-	/* XPS mapping queues to 0..N cpus (may be less than ntxqs) */
-	for_each_online_cpu(cpu)
-		netif_set_xps_queue(port->dev, cpumask_of(cpu), cpu);
-
 	if (port->has_tx_irqs) {
 		/* Download time-coal. The pkts-coal done in start_dev */
 		mvpp2_tx_time_coal_set(port);
@@ -5710,6 +5702,10 @@ static int mvpp2_port_probe(struct platform_device *pdev,
 	if (!dev)
 		return -ENOMEM;
 
+	/* XPS mapping queues to 0..N cpus (may be less than ntxqs) */
+	for_each_online_cpu(cpu)
+		netif_set_xps_queue(dev, cpumask_of(cpu), cpu);
+
 	phy_mode = fwnode_get_phy_mode(port_fwnode);
 	if (phy_mode < 0) {
 		dev_err(&pdev->dev, "incorrect phy mode\n");
-- 
2.17.1

